### PR TITLE
APPS-994 Fix bug in frag applet source code, affecting applet re-use

### PR DESCRIPTION
**Pull request by r-i-v-a**

### CODE REVIEW

File: compiler/src/main/scala/dx/translator/cwl/ProcessTranslator.scala [hunk 1]
```diff
@@ -476,6 +476,8 @@ case class ProcessTranslator(cwlBundle: CwlBundle,
       Stage(call.name, getStage(), calleeName, inputs, callee.outputVars)
     }
 
+    // TODO this is where fragment applet is built
```
Comment (r-i-v-a) [inline compiler/src/main/scala/dx/translator/cwl/ProcessTranslator.scala]: For WDL, it's here
compiler/src/main/scala/dx/translator/wdl/CallableTranslator.scala

File: compiler/src/main/scala/dx/translator/wdl/CallableTranslator.scala [hunk 2]
```diff
@@ -526,6 +527,8 @@ case class CallableTranslator(wdlBundle: WdlBundle,
                 |""".stripMargin
         )
         val blockName = s"${wfName}_block_${pathStr}"
+
+        // TODO block subworkflow is created here
```
Comment (commandlinegirl) [inline compiler/src/main/scala/dx/translator/wdl/CallableTranslator.scala]: Please update the comment to sth like: A block subworkflow is created here

File: core/src/main/scala/dx/core/ir/package.scala [hunk 3]
```diff
@@ -146,6 +146,9 @@ object ExecutableKind {
   */
 trait RuntimeRequirement
 
+// TODO case class Application defined here
```
Comment (commandlinegirl) [inline core/src/main/scala/dx/core/ir/package.scala]: Can this be removed?

**Review by commandlinegirl — CHANGES_REQUESTED** [review]

File: core/src/main/scala/dx/core/languages/wdl/WdlCodeGenerator.scala [hunk 4]
```diff
@@ -286,6 +294,26 @@ case class CodeGenerator(typeAliases: Map[String, WdlTypes.T_Struct],
...
+        comments = CommentMap.empty
+    )(SourceLocation.empty)
+  }
+
+  def createWdlBlockSrc(subBlocks: Vector[WdlBlock]): TAT.Document = {
+    TAT.Document(
+        source = StringFileNode(contents = subBlocks.map(_.prettyFormat).mkString),
+        version = TAT.Version(outputWdlVersion)(SourceLocation.empty),
+        elements = structDefs,
+        workflow = None,
```
Comment (commandlinegirl) [inline core/src/main/scala/dx/core/languages/wdl/WdlCodeGenerator.scala]: Should this be always None? What about a case when the block is a subworkflow?

**Review by commandlinegirl — CHANGES_REQUESTED** [review]

File: core/src/main/scala/dx/core/languages/wdl/WdlCodeGenerator.scala (hunk 4 above)
Comment (Gvaihir) [reply core/src/main/scala/dx/core/languages/wdl/WdlCodeGenerator.scala]: AFAIK, blocks are interpreted as workflow objects on the platform anyway. However if in the `.wdl` there's an actual workflow nested in the main workflow (marked by the word `workflow` - those are parsed as workflows from the very beginning along with tasks. At the point where dependency graph is created. But I may be wrong. This code generator is acting at a much later stage, when dependency graph is constructed.

Comment (Gvaihir) [reply core/src/main/scala/dx/core/languages/wdl/WdlCodeGenerator.scala]: I retract that, fixing

File: core/src/main/scala/dx/core/ir/package.scala (hunk 3 above)
Comment (Gvaihir) [reply core/src/main/scala/dx/core/ir/package.scala]: Fixed in [9c3ecae](https://github.com/dnanexus/dxCompiler/pull/291/commits/9c3ecae9726a388bdd2185c61bc605de4f5f251f)

File: compiler/src/main/scala/dx/translator/wdl/CallableTranslator.scala (hunk 2 above)
Comment (Gvaihir) [reply compiler/src/main/scala/dx/translator/wdl/CallableTranslator.scala]: Fixed in [9c3ecae](https://github.com/dnanexus/dxCompiler/pull/291/commits/9c3ecae9726a388bdd2185c61bc605de4f5f251f)

File: core/src/main/scala/dx/core/languages/wdl/WdlCodeGenerator.scala (hunk 4 above)
Comment (Gvaihir) [reply core/src/main/scala/dx/core/languages/wdl/WdlCodeGenerator.scala]: Attempt to fix in [9c3ecae](https://github.com/dnanexus/dxCompiler/pull/291/commits/9c3ecae9726a388bdd2185c61bc605de4f5f251f)
Checking

Comment (Gvaihir) [reply core/src/main/scala/dx/core/languages/wdl/WdlCodeGenerator.scala]: Fixed

**Comment by Gvaihir** [comment]
Unit tests fail because of the dependency on [wdlTools](https://github.com/dnanexus/wdlTools/pull/213) - pls approve that first

File: RELEASE_NOTES.md [hunk 5]
```diff
@@ -2,7 +2,9 @@
 
 ## in develop
 
-* ..
+* WDL: Stages (frags/blocks) reuse applets as well. Previously only tasks were reused. **Breaking**: can break the logic
```
Comment (commandlinegirl) [inline RELEASE_NOTES.md]: How about we change it to: Fragments and blocks reuse applets as well, i.e. they are not rebuilt if the code corresponding to them hasn't been updated in the WDL source file. Previously only tasks..

File: RELEASE_NOTES.md [hunk 6]
```diff
@@ -2,7 +2,9 @@
...
+  of any App reuse for CWL (even tasks maybe are not reused), because `ApplicationCompiler` and `WorkflowCompiler` now
+  look at the `DocContents` for checksum, and `SourceCode` attribute is now ignored.
```
Comment (commandlinegirl) [inline RELEASE_NOTES.md]: Do we need a separate CWL ticket to test this?

File: compiler/src/test/resources/bugs/apps_994_v2.wdl [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+	output {
+		File out = "reuse_print.txt"
+	}
+}
+
+workflow reuse {
+
+	# Frag applet via sub-expression
+	# declaration of inputs is dofferent from V1. Therefore this frag should be rebuilt
```
Comment (commandlinegirl) [inline compiler/src/test/resources/bugs/apps_994_v2.wdl]: *different

**Review by commandlinegirl — DISMISSED** [review]
LGTM modulo minor bits.

**Comment by commandlinegirl** [comment]
@Gvaihir I've set the `cwl_conformance` label as well.

File: RELEASE_NOTES.md (hunk 6 above)
Comment (Gvaihir) [reply RELEASE_NOTES.md]: I'd fix that in a separate ticket - otherwise APPS-994 gets real big and with lots of comments

Comment (commandlinegirl) [reply RELEASE_NOTES.md]: That's reasonable. If you could add a note to `CWL only` https://github.com/dnanexus/dxCompiler#limitations about a limited functionality for applet reuse for CWL. Thanks!

File: RELEASE_NOTES.md (hunk 5 above)
Comment (Gvaihir) [reply RELEASE_NOTES.md]: fixed in [98647f0](https://github.com/dnanexus/dxCompiler/pull/291/commits/98647f0cb1881747656c96096189f220c7a0b1d6)

File: compiler/src/test/resources/bugs/apps_994_v2.wdl (hunk 7 above)
Comment (Gvaihir) [reply compiler/src/test/resources/bugs/apps_994_v2.wdl]: fixed in [98647f0](https://github.com/dnanexus/dxCompiler/pull/291/commits/98647f0cb1881747656c96096189f220c7a0b1d6)

File: RELEASE_NOTES.md (hunk 6 above)
Comment (Gvaihir) [reply RELEASE_NOTES.md]: added in [ab5d21b](https://github.com/dnanexus/dxCompiler/pull/291/commits/ab5d21bf743f7c619179f0c47c15839b98f53ec6)

**Review by commandlinegirl — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**